substratevm/src/com.oracle.svm.core.genscavenge/src/com/oracle/svm/core/genscavenge/parallel/ParallelGCImpl.java

```diff
@@ -0,0 +1,258 @@
+import com.oracle.svm.core.heap.ParallelGC;
+import com.oracle.svm.core.jdk.Jvm;
+import com.oracle.svm.core.locks.VMCondition;
+import com.oracle.svm.core.locks.VMMutex;
+import com.oracle.svm.core.log.Log;
+import com.oracle.svm.core.thread.VMThreads;
+import com.oracle.svm.core.threadlocal.FastThreadLocalFactory;
+import com.oracle.svm.core.threadlocal.FastThreadLocalWord;
+import com.oracle.svm.core.util.VMError;
+import org.graalvm.compiler.api.replacements.Fold;
+import org.graalvm.nativeimage.ImageSingletons;
+import org.graalvm.nativeimage.Platform;
+import org.graalvm.nativeimage.Platforms;
+import org.graalvm.word.Pointer;
+import org.graalvm.word.WordFactory;
+
+import java.util.concurrent.atomic.AtomicInteger;
+
+public class ParallelGCImpl extends ParallelGC {
+
+    public static final int UNALIGNED_BIT = 0x01;
+    private static final int MAX_WORKERS_COUNT = 31;
+
+    /**
+     * Number of parallel GC workers. Default is 1 so that if GC happens before worker threads are started,
+     * it occurs on a single thread just like serial GC.
+     */
+    private int workerCount = 1;
+    private int allWorkersBusy;
+    private final AtomicInteger busyWorkers = new AtomicInteger(0);
+
+    /**
+     * Each GC worker allocates memory in its own thread local chunk, entering mutex only when new chunk needs to be allocated.
+     */
+    private static final FastThreadLocalWord<AlignedHeapChunk.AlignedHeader> allocChunkTL =
+            FastThreadLocalFactory.createWord("ParallelGCImpl.allocChunkTL");
+    private static final FastThreadLocalWord<AlignedHeapChunk.AlignedHeader> scannedChunkTL =
+            FastThreadLocalFactory.createWord("ParallelGCImpl.scannedChunkTL");
+
+    public static final VMMutex mutex = new VMMutex("ParallelGCImpl");
+    private final VMCondition seqPhase = new VMCondition(mutex);
+    private final VMCondition parPhase = new VMCondition(mutex);
+
+    private ChunkBuffer buffer;
+    private volatile boolean inParallelPhase;
+
+    @Fold
+    public static ParallelGCImpl singleton() {
+        return (ParallelGCImpl) ImageSingletons.lookup(ParallelGC.class);
+    }
+
+    public static boolean isInParallelPhase() {
+        return singleton().inParallelPhase;
+    }
+
+    public static void waitForIdle() {
+        singleton().waitForIdleImpl();
+    }
+
+    public static AlignedHeapChunk.AlignedHeader getThreadLocalScannedChunk() {
+        return scannedChunkTL.get();
+    }
+
+    public static AlignedHeapChunk.AlignedHeader getThreadLocalChunk() {
+        return allocChunkTL.get();
+    }
+
+    public static void setThreadLocalChunk(AlignedHeapChunk.AlignedHeader chunk) {
+        allocChunkTL.set(chunk);
+    }
+
+    public void push(Pointer ptr) {
+        buffer.push(ptr);
+    }
+
+    @Override
+    public void startWorkerThreadsImpl() {
+        int hubOffset = ConfigurationValues.getObjectLayout().getHubOffset();
+        VMError.guarantee(hubOffset == 0, "hub offset must be 0");
```

Comment:
Because the object copying code in its current form -- copy header, then the rest of the bits -- is only valid if `hubOffset` is 0, otherwise I need to additionally copy the `[0, hubOffset)` part. From the code I was assuming hub offset is always zero, but it might be worthwhile to add that missing part just for consistency. Since `getHubOffset()` is folded, that part should be removed at compile time anyway